File: Unity/GeometrySync/Assets/GeometrySync/Runtime/MeshDeserializer.cs
using System;
using UnityEngine;

namespace GeometrySync
{
    /// <summary>
    /// Mesh data container
    /// </summary>
    public struct MeshData
    {
        public Vector3[] Vertices;
        public Vector3[] Normals;
        public Vector2[] UVs;
        public int[] Indices;

        public int VertexCount => Vertices?.Length ?? 0;
        public int TriangleCount => (Indices?.Length ?? 0) / 3;
    }

    /// <summary>
    /// Instance data container for GPU instancing
    /// </summary>
    public struct InstanceData
    {
        public uint MeshId;              // Hash of base mesh name
        public Matrix4x4[] Transforms;   // Instance transform matrices

        public int InstanceCount => Transforms?.Length ?? 0;
    }

    /// <summary>
    /// Deserializes binary mesh data from Blender
    /// </summary>
    public static class MeshDeserializer
    {
        /// <summary>
        /// Deserialize mesh data from binary format
        ///
        /// Binary format:
        /// - Header: vertex_count (uint32), index_count (uint32)
        /// - Vertex data: interleaved [x,y,z, nx,ny,nz, u,v] as float32 (32 bytes per vertex)
        /// - Index data: uint32 array
        /// </summary>
        public static MeshData Deserialize(byte[] data)
        {
            if (data == null || data.Length < 8)
            {
                throw new ArgumentException("Invalid mesh data: too small");
            }

            int offset = 0;

            // Read header
            uint vertexCount = BitConverter.ToUInt32(data, offset);
            offset += 4;

            uint indexCount = BitConverter.ToUInt32(data, offset);
            offset += 4;

            // Validate counts
            if (vertexCount > 10_000_000 || indexCount > 30_000_000)
            {
                throw new ArgumentException($"Mesh too large: {vertexCount} vertices, {indexCount} indices");
            }

            int expectedVertexDataSize = (int)vertexCount * 32; // 8 floats * 4 bytes
            int expectedIndexDataSize = (int)indexCount * 4;    // uint32
            int expectedTotalSize = 8 + expectedVertexDataSize + expectedIndexDataSize;

            if (data.Length < expectedTotalSize)
            {
                throw new ArgumentException($"Invalid mesh data size. Expected {expectedTotalSize}, got {data.Length}");
            }

            // Allocate arrays
            Vector3[] vertices = new Vector3[vertexCount];
            Vector3[] normals = new Vector3[vertexCount];
            Vector2[] uvs = new Vector2[vertexCount];
            int[] indices = new int[indexCount];

            // Read interleaved vertex data
            for (int i = 0; i < vertexCount; i++)
            {
                // Position (3 floats)
                float x = BitConverter.ToSingle(data, offset);
                offset += 4;
                float y = BitConverter.ToSingle(data, offset);
                offset += 4;
                float z = BitConverter.ToSingle(data, offset);
                offset += 4;
                vertices[i] = new Vector3(x, y, z);

                // Debug: print first vertex
                if (i == 0)
                {
                    UnityEngine.Debug.Log($"[Deserializer] First vertex: ({x}, {y}, {z})");
                }

                // Normal (3 floats)
                float nx = BitConverter.ToSingle(data, offset);
                offset += 4;
                float ny = BitConverter.ToSingle(data, offset);
                offset += 4;
                float nz = BitConverter.ToSingle(data, offset);
                offset += 4;
                normals[i] = new Vector3(nx, ny, nz);

                // UV (2 floats)
                float u = BitConverter.ToSingle(data, offset);
                offset += 4;
                float v = BitConverter.ToSingle(data, offset);
                offset += 4;
                uvs[i] = new Vector2(u, v);
            }

            // Read indices
            for (int i = 0; i < indexCount; i++)
            {
                indices[i] = (int)BitConverter.ToUInt32(data, offset);
                offset += 4;
            }

            return new MeshData
            {
                Vertices = vertices,
                Normals = normals,
                UVs = uvs,
                Indices = indices
            };
        }

        /// <summary>
        /// Deserialize using Unity's NativeArray for zero-copy (future optimization)
        /// Note: This is a placeholder for Phase 3 optimization
        /// </summary>
        public static MeshData DeserializeNative(byte[] data)
        {
            // This will be implemented in Phase 3 for better performance
            // Currently just calls the standard Deserialize method
            return Deserialize(data);
        }

        /// <summary>
        /// Deserialize instance data from binary format (Phase 2: GPU Instancing)
        ///
        /// Binary format:
        /// - Header: mesh_id (uint32), instance_count (uint32)
        /// - Transform data: array of 4x4 matrices (float32, 16 values per matrix)
        /// </summary>
        public static InstanceData DeserializeInstanceData(byte[] data)
        {
            if (data == null || data.Length < 8)
            {
                throw new ArgumentException("Invalid instance data: too small");
            }

            int offset = 0;

            // Read header
            uint meshId = BitConverter.ToUInt32(data, offset);
            offset += 4;

            uint instanceCount = BitConverter.ToUInt32(data, offset);
            offset += 4;

            // Sanity check: prevent memory issues
            if (instanceCount > 100_000)
            {
                throw new ArgumentException($"Instance count too large: {instanceCount} (max 100,000)");
            }

            // Expected data size: 8 (header) + instanceCount * 16 * 4 (64 bytes per matrix)
            int expectedSize = 8 + (int)instanceCount * 64;
            if (data.Length < expectedSize)
            {
                throw new ArgumentException(
                    $"Invalid instance data size: expected {expectedSize}, got {data.Length}");
            }

            // Read transform matrices (16 floats per matrix)
            Matrix4x4[] transforms = new Matrix4x4[instanceCount];

            for (int i = 0; i < instanceCount; i++)
            {
                // Read 16 floats for 4x4 matrix
                float[] m = new float[16];
                for (int j = 0; j < 16; j++)
                {
                    m[j] = BitConverter.ToSingle(data, offset);
                    offset += 4;
                }

                // Construct Matrix4x4 from column-major data
                // Unity uses column-major order: Matrix4x4(column0, column1, column2, column3)
                transforms[i] = new Matrix4x4(
                    new Vector4(m[0], m[1], m[2], m[3]),      // column 0
                    new Vector4(m[4], m[5], m[6], m[7]),      // column 1
                    new Vector4(m[8], m[9], m[10], m[11]),    // column 2
                    new Vector4(m[12], m[13], m[14], m[15])   // column 3
                );
            }

            return new InstanceData
            {
                MeshId = meshId,
                Transforms = transforms
            };
        }
    }
}

File: VisionPro/GeometrySyncVisionPro/GeometrySyncVisionPro/Protocol/MeshDeserializer.swift
import Foundation
import simd

/// Deserializes binary mesh data from Blender.
/// Port of Unity/GeometrySync/.../MeshDeserializer.cs
enum MeshDeserializer {

    // MARK: - Validation limits

    static let maxVertexCount: UInt32 = 10_000_000
    static let maxIndexCount: UInt32 = 30_000_000
    static let maxInstanceCount: UInt32 = 100_000

    // MARK: - Mesh deserialization

    /// Validate the mesh header and return zero-copy slices of the payload.
    ///
    /// Binary format:
    /// - Header: vertexCount (uint32 LE), indexCount (uint32 LE)
    /// - Vertex data: interleaved [x,y,z, nx,ny,nz, u,v] as float32 LE (32 bytes per vertex)
    /// - Index data: uint32 LE array, wire winding (i0, i1, i2)
    ///
    /// This is the fast path for the Metal renderer: no per-vertex parsing —
    /// the slices are uploaded to GPU staging buffers as-is and unpacked by
    /// compute kernels (including the RealityKit winding reversal).
    static func rawMesh(from data: Data) throws -> RawMeshData {
        guard data.count >= 8 else {
            throw DeserializerError.dataTooSmall(data.count)
        }

        let (vertexCount, indexCount) = data.withUnsafeBytes { raw -> (UInt32, UInt32) in
            let ptr = raw.baseAddress!
            var offset = 0
            let v = readUInt32(ptr, offset: &offset)
            let i = readUInt32(ptr, offset: &offset)
            return (v, i)
        }

        guard vertexCount <= maxVertexCount, indexCount <= maxIndexCount else {
            throw DeserializerError.meshTooLarge(
                vertices: Int(vertexCount), indices: Int(indexCount))
        }

        let vertexDataSize = Int(vertexCount) * RawMeshData.vertexStride
        let indexDataSize = Int(indexCount) * 4
        let expectedTotalSize = 8 + vertexDataSize + indexDataSize

        guard data.count >= expectedTotalSize else {
            throw DeserializerError.invalidDataSize(
                expected: expectedTotalSize, got: data.count)
        }

        let vertexStart = data.startIndex + 8
        let indexStart = vertexStart + vertexDataSize

        // Slices, not copies — they retain the payload's backing storage,
        // which goes straight into Metal staging buffers.
        return RawMeshData(
            vertexCount: Int(vertexCount),
            indexCount: Int(indexCount),
            vertexData: data[vertexStart..<indexStart],
            indexData: data[indexStart..<(indexStart + indexDataSize)]
        )
    }

    /// CPU fallback: expand a validated raw payload into parsed arrays.
    ///
    /// Applies the winding order reversal for RealityKit (right-hand system):
    /// Blender→Unity produces left-hand winding (i0, i1, i2),
    /// RealityKit needs right-hand winding: (i0, i2, i1).
    static func expand(_ raw: RawMeshData) -> MeshData {
        let vCount = raw.vertexCount
        let iCount = raw.indexCount

        var positions = [SIMD3<Float>]()
        positions.reserveCapacity(vCount)
        var normals = [SIMD3<Float>]()
        normals.reserveCapacity(vCount)
        var uvs = [SIMD2<Float>]()
        uvs.reserveCapacity(vCount)

        raw.vertexData.withUnsafeBytes { rawBytes in
            let ptr = rawBytes.baseAddress!
            var offset = 0
            for _ in 0..<vCount {
                let x = readFloat(ptr, offset: &offset)
                let y = readFloat(ptr, offset: &offset)
                let z = readFloat(ptr, offset: &offset)
                positions.append(SIMD3<Float>(x, y, z))

                let nx = readFloat(ptr, offset: &offset)
                let ny = readFloat(ptr, offset: &offset)
                let nz = readFloat(ptr, offset: &offset)
                normals.append(SIMD3<Float>(nx, ny, nz))

                let u = readFloat(ptr, offset: &offset)
                let v = readFloat(ptr, offset: &offset)
                uvs.append(SIMD2<Float>(u, v))
            }
        }

        var indices = [UInt32](repeating: 0, count: iCount)
        raw.indexData.withUnsafeBytes { rawBytes in
            let ptr = rawBytes.baseAddress!
            var offset = 0
            let triangleCount = iCount / 3
            for tri in 0..<triangleCount {
                let baseIdx = tri * 3
                let i0 = readUInt32(ptr, offset: &offset)
                let i1 = readUInt32(ptr, offset: &offset)
                let i2 = readUInt32(ptr, offset: &offset)
                indices[baseIdx]     = i0
                indices[baseIdx + 1] = i2  // swapped
                indices[baseIdx + 2] = i1  // swapped
            }
            // Handle remaining indices (if indexCount is not a multiple of 3)
            for i in (triangleCount * 3)..<iCount {
                indices[i] = readUInt32(ptr, offset: &offset)
            }
        }

        return MeshData(
            positions: positions,
            normals: normals,
            uvs: uvs,
            indices: indices
        )
    }

    /// Deserialize mesh data from binary format (validation + full CPU parse).
    static func deserializeMesh(_ data: Data) throws -> MeshData {
        expand(try rawMesh(from: data))
    }

    // MARK: - Instance deserialization

    /// Deserialize instance data from binary format.
    ///
    /// Binary format:
    /// - Header: meshId (uint32 LE), instanceCount (uint32 LE)
    /// - Transform data: array of 4x4 matrices (16 × float32 LE per matrix, column-major)
    static func deserializeInstances(_ data: Data) throws -> InstanceData {
        guard data.count >= 8 else {
            throw DeserializerError.dataTooSmall(data.count)
        }

        return try data.withUnsafeBytes { raw in
            let ptr = raw.baseAddress!
            var offset = 0

            let meshId = readUInt32(ptr, offset: &offset)
            let instanceCount = readUInt32(ptr, offset: &offset)

            guard instanceCount <= maxInstanceCount else {
                throw DeserializerError.instanceCountTooLarge(Int(instanceCount))
            }

            let expectedSize = 8 + Int(instanceCount) * 64
            guard data.count >= expectedSize else {
                throw DeserializerError.invalidDataSize(
                    expected: expectedSize, got: data.count)
            }

            var transforms = [simd_float4x4]()
            transforms.reserveCapacity(Int(instanceCount))

            for _ in 0..<instanceCount {
                // Read 16 floats for 4×4 column-major matrix
                let col0 = SIMD4<Float>(
                    readFloat(ptr, offset: &offset),
                    readFloat(ptr, offset: &offset),
                    readFloat(ptr, offset: &offset),
                    readFloat(ptr, offset: &offset)
                )
                let col1 = SIMD4<Float>(
                    readFloat(ptr, offset: &offset),
                    readFloat(ptr, offset: &offset),
                    readFloat(ptr, offset: &offset),
                    readFloat(ptr, offset: &offset)
                )
                let col2 = SIMD4<Float>(
                    readFloat(ptr, offset: &offset),
                    readFloat(ptr, offset: &offset),
                    readFloat(ptr, offset: &offset),
                    readFloat(ptr, offset: &offset)
                )
                let col3 = SIMD4<Float>(
                    readFloat(ptr, offset: &offset),
                    readFloat(ptr, offset: &offset),
                    readFloat(ptr, offset: &offset),
                    readFloat(ptr, offset: &offset)
                )
                transforms.append(simd_float4x4(col0, col1, col2, col3))
            }

            return InstanceData(meshId: meshId, transforms: transforms)
        }
    }

    // MARK: - Binary readers (little-endian)

    @inline(__always)
    private static func readUInt32(_ base: UnsafeRawPointer, offset: inout Int) -> UInt32 {
        let value = base.loadUnaligned(fromByteOffset: offset, as: UInt32.self)
        offset += 4
        return UInt32(littleEndian: value)
    }

    @inline(__always)
    private static func readFloat(_ base: UnsafeRawPointer, offset: inout Int) -> Float {
        let bits = base.loadUnaligned(fromByteOffset: offset, as: UInt32.self)
        offset += 4
        return Float(bitPattern: UInt32(littleEndian: bits))
    }
}

// MARK: - Errors

enum DeserializerError: LocalizedError {
    case dataTooSmall(Int)
    case meshTooLarge(vertices: Int, indices: Int)
    case instanceCountTooLarge(Int)
    case invalidDataSize(expected: Int, got: Int)

    var errorDescription: String? {
        switch self {
        case .dataTooSmall(let size):
            return "Invalid data: too small (\(size) bytes)"
        case .meshTooLarge(let v, let i):
            return "Mesh too large: \(v) vertices, \(i) indices"
        case .instanceCountTooLarge(let count):
            return "Instance count too large: \(count) (max \(MeshDeserializer.maxInstanceCount))"
        case .invalidDataSize(let expected, let got):
            return "Invalid data size: expected \(expected), got \(got)"
        }
    }
}
